libs/@local/harpc/crates/wire/src/request/mod.rs

```diff
@@ -0,0 +1,327 @@
+pub use bytes::Bytes;
+use error_stack::{Result, ResultExt};
+use tokio::io::{AsyncRead, AsyncWrite};
+
+use self::{
+    body::{RequestBody, RequestBodyContext},
+    codec::{DecodeError, EncodeError},
+    header::RequestHeader,
+};
+use crate::codec::{Decode, DecodePure, Encode};
+
+pub mod authorization;
+pub mod begin;
+pub mod body;
+pub mod codec;
+pub mod encoding;
+pub mod flags;
+pub mod frame;
+pub mod header;
+pub mod id;
+pub mod procedure;
+pub mod service;
+
+/// A request message.
+///
+/// A request message is composed of a header and a body and is used to send the actual message over
+/// the wire, agnostic over the transport layer.
+///
+/// This is a binary protocol with minimal overhead with framing support. Messages can be up to
+/// 64KiB in size, larger messages are split into multiple frames.
+///
+/// The transport layer is responsible for ensuring that the message is sent and received in order.
+///
+///
+/// # `Begin` Packet
+///
+/// The layout of a `Begin` packet is as follows:
+///
+/// ```text
+/// 0                   1                   2                   3
+/// 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1
+/// +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+/// |  Magic  |P|R. |F|S. |S. |P. |E. |A. |      Authorization      |
+/// +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+/// |     |P. |                       Payload                       |
+/// +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+///
+/// * Magic (5 bytes)
+/// * Protocol Version (1 byte)
+/// * Request Id (2 bytes)
+/// * Flags (1 byte)
+/// * Service Id (2 bytes)
+/// * Service Version (2 bytes)
+/// * Procedure Id (2 bytes)
+/// * Encoding (2 bytes)
+/// * Accept (2 bytes)
+/// * Authorization (16 bytes)
```

Comment:
Worth flagging that authorization may not fit into 16 bytes in the future. We can assume this for now but I'm not sure if we can keep this constraint. However, it could be possible we will continue using a single UUID but not the account ID but a generated token.

Reply:
I have been thinking of moving authorization entirely out of the header instead of relying on the new session mechanism. Basically, you call `0xF0`, and `0x00` to authenticate (`0xFx` is reserved for internal communication) to authenticate a session, and then that's that, instead of needing an authentication on every request. This would also make the packet smaller, albeit increase implementation complexity for authentication.

Reply:
I think it's fine to move authorization out. I'm not sure about authentication but I think it shouldn't be in the header as well as the size may vary between applications (and it's wasted space if not required). We should chose the more performant way, adding a bit of complexity to the client is fine IMO.

Reply:
sounds good, in that case we'll designate a service in the reserved `0xF` range for authentication and authorization during a session 👍 (it also brings down the header size which is always nice)

Reply:
Just to understand: Authorization in this context means "The user is allowed to call the endpoint at all", right?

Reply:
Yes, the idea is simple, in more conceptual terms you call:

```
authenticate(accountId: Uuid);
```

At the beginning of the session. Then every request after that is associated with that account (calling `authenticate` multiple times in the same session or not at the start will result in an error).

Endpoints then *authorize* against the account we provided in the session, denying or allowing the request just like we're doing right now. So nothing really changed there, if that makes sense.

Reply:
Currently, we invoke an external service (SpiceDB) and decide on a per-request basis if something is allowed or not. This highly depends on the parameters passed in (A user might update entity A but not entity B). Also, account ID is only used because the graph is not exposed to user-facing code (the frontend cannot file requests) but it's abstracted away. Otherwise it would be trivial to use a more powerful account ID. At some point this needs to be changed so a more opaque ID is used (e.g. a token).

If this is moved to the `0xF000` interface I guess the size of the authentication parameters would be more dynamic (similar to the payload?), so we would not limit ourselves to 16 bytes?

Reply:
Yes, this would still apply, the code for authorization would be still on per RPC basis, but the idea is that the `accountId` is the same (or an opaque token) across all RPC method calls in that particular session (see the SVG).

it boils down to:


```
pub struct Authentication {
	account: AccountId
}

fn authenticate(session: &Session, authorization: Authorization) -> Result<(), AuthenticationError> {
	// ... e.g. pre-fetch all permissions, or simply do:
	session.insert(account);
}

// then in a "normal" RPC method:

fn do_thing(session: &Session, args: Args) -> Result<(), ThingError> {
	let account = session.get_or_err(account).change_context(ThingError)?;

	// ... check that the user is authorized
	if !account_is_authorized(account) {
		return Err(Report::new(Unauthorized(account)));
	}

	Ok(());
}
```


The key is that this is per session, a longer-lived connection used to do multiple requests with a single user. If a user changes, one creates a new session (a new connection).